scripts/import-cp-contexts.sh

```diff
@@ -0,0 +1,249 @@
+--merge                                 merge the control planes contexts with the existing cluster contexts
+-o <filename>                           specify a different kubeconfig file to save the contexts (- for stdout)
+--silent                                no information output
+-X                                      enable verbose execution of the script for debugging
+EOF
+}
+
+
+get_kubeconfig() {
+    context="$1"
+    cp_name="$2"
+    cp_type="$3"
+    ip_addr="$4"
+
+    echov "Getting the kubeconfig of Control Plane \"$cp_name\" of type \"$cp_type\" in context \"$context\":"
+
+    # wait for CP ready
+    while [[ $(KUBECONFIG="$in_KUBECONFIG" kubectl --context $context get controlplane $cp_name -o 'jsonpath={..status.conditions[?(@.type=="Ready")].status}') != "True" ]]; do
+        echov "* Waiting for Control Plane \"$cp_name\" in context \"$context\" to be Ready..."
+        sleep 5
+    done
+
+    # get out of cluster kubeconfig
+    if [[ "$cp_type" == "host" ]] ; then
+        echov "- Using cluster context \"${context}\""
+        kubeconfig=$(KUBECONFIG="$in_KUBECONFIG" kubectl --context $context config view --flatten --minify)
+    else
+        # determine the secret name and namespace
+                    key=$(KUBECONFIG="$in_KUBECONFIG" kubectl --context $context get controlplane $cp_name -o=jsonpath='{.status.secretRef.key}')
+            secret_name=$(KUBECONFIG="$in_KUBECONFIG" kubectl --context $context get controlplane $cp_name -o=jsonpath='{.status.secretRef.name}')
+        secret_namespace=$(KUBECONFIG="$in_KUBECONFIG" kubectl --context $context get controlplane $cp_name -o=jsonpath='{.status.secretRef.namespace}')
+        # get the kubeconfig
+        echov "- Using \"$key\" from \"$secret_name\" secret in \"$secret_namespace\""
+        kubeconfig=$(KUBECONFIG="$in_KUBECONFIG" kubectl --context $context get secret $secret_name -n $secret_namespace -o=jsonpath="{.data.$key}" | base64 -d)
+    fi
+
+    # make kubeconfig unique for the control plane
+    echov "- Making the kubeconfig unique..."
+    cluster=$(kubectl --kubeconfig <(echo "$kubeconfig") config get-clusters | tail -n +2)
+    user=$(kubectl --kubeconfig <(echo "$kubeconfig") config get-users | tail -n +2)
+    kubeconfig=$(
+        echo "$kubeconfig" \
+        | yq ".clusters[0].name = \"$cp_name-cluster\"" \
+        | yq ".users[0].name = \"$cp_name-user\"" \
+        | yq ".contexts[0].name = \"$cp_name\"" \
+        | yq ".contexts[0].context.cluster = \"$cp_name-cluster\"" \
+        | yq ".contexts[0].context.user = \"$cp_name-user\"" \
+        | yq ".current-context = \"$cp_name\""
+    )
+
+    # swap out 127.0.0.1 with an external ip address
+    if [[ "$ip_addr" != "" ]] ; then
+        echov "- Replacing server ip address \"127.0.0.1\" with \"$ip_addr\""
+        kubeconfig=$(echo "$kubeconfig" | sed "s@server: https://127\.0\.0\.1:@server: https://$ip_addr:@g")
```

Comment:
Looking in my secrets, I do not find "127.0.0.1":

```
mspreitz@mjs13 kubestellar % kubectl get secrets vc-vcluster -n its1-system -o jsonpath='{.data.config}' | base64 -d | fgrep server:
    server: https://its1.localtest.me:9443
```

Reply:
```
mspreitz@mjs13 kubestellar % kubectl get secrets admin-kubeconfig -n wds1-system -o jsonpath='{.data.kubeconfig}' | base64 -d | fgrep server:
    server: https://wds1.localtest.me:9443
```

Reply:
@MikeSpreitzer the thought here is to aid in reaching certain CPs from outside the host of the core cluster... i.e. WECs running in a different machine/VM. More specifically in the case of kind/k3s. Now, so far the only way to do this has been to use CPs of type host, since their kubeconfig == cluster kubeconfig, it may have a cluster 127.0.0.1, which cannot be reached externally... so for this narrow case a replacement can be useful.

None of this applies to OCP that will create fully routable domains for its routes.

In conclusions:

for OCP clusters, the argument is not needed, not useful, will not do anything

for kind/k3s clusters with CPs of type <> host, the argument is not needed, not useful, will not do anything ==> the CP will not  be accessible from outside the host machine  (== we have not found a way yet)

Reply:
I think this will become more clear when we work on practical examples of usage of the script... like for the bare metal RPI case

Reply:
OK. This should be documented somewhere.

Reply:
Also, I just realized: part of the problem is that when I read "local-ip-address" in the flag name, I did not understand that to mean '127.0.0.1' --- I thought it meant some IP address that the local machine has. Perhaps changing the flag name to something like `--replace-localhost` or `--replace-127001` (or `--replace-127-0-0-1` if you prefer).

Reply:
ok, I changed to `--replace-localhost `